UCT/CUDA/CUDA_COPY: Changed log level in mem_free.

rakhmets opened the pull request:
## What?
Changed log level used in `uct_cuda_copy_mem_free`. `DIAG` log level is enough for the purpose of the function.

## Why?
The memory can be implicitly destroyed using CUDA Driver/Runtime API: e.g. `cuDevicePrimaryCtxRelease`, `cuDevicePrimaryCtxReset`, `cudaDeviceReset`. The following error may be observed during the resources clean up:
```
cuda_copy_md.c:610  UCX  ERROR cuMemFree_v2(alloc_handle->ptr) failed: invalid argument
      ucp_mm.c:460  UCX  WARN  failed to free: Input/output error
```

yosefe commented on src/uct/cuda/cuda_copy/cuda_copy_md.c:
```diff
@@ -616,7 +616,8 @@ static ucs_status_t uct_cuda_copy_mem_free(uct_md_h md, uct_mem_h memh)
         status = uct_cuda_copy_mem_release_fabric(alloc_handle);
     } else {
-        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemFree(alloc_handle->ptr));
+        UCT_CUDADRV_FUNC_LOG_DEBUG(cuMemFree(alloc_handle->ptr));
```
IMO it should be UCT_CUDADRV_FUNC_LOG_WARN

rakhmets replied:
If the log level is `WARNING`, we will print a warning in the following scenario:
```C
void cuda_set_device()
{
    cudaSetDevice();
    /**
      * , or 
      * cuDevicePrimaryCtxRetain();
      * cuCtxPushCurrent();
      * , or 
      * cuCtxCreate();
      */
}

void cuda_reset_device()
{
    cudaDeviceReset();
    /**
     * , or
     * cuCtxPopCurrent();
     * cuDevicePrimaryCtxRelease();
     * , or
     * cuCtxDestroy();
     */
}

int main()
{
    ...
    MPI_Init();
    ...
    cuda_set_device();
    ...
    /* UCX creates staging buffer using cuMemAlloc */
    ...
    cuda_reset_device();
    ...
    MPI_Fnalize();
    ...
}
```
Do we need a warning here?

P.S. We can add TODO here, to add CUDA context Id to `uct_cuda_copy_alloc_handle_t`, so we can check if the context is still alive.

yosefe replied:
yes, let's do this, to still have warning if cuMemFree failed for another reason

rakhmets replied:
Force-pushed changes since it a one-line change.

brminich replied:
what does it fix then? We will still have warning in the cases you mentioned. What does returning UCS_OK fix?

brminich reviewed (changes requested):
imo, this change does not make sence
if we use debug log returning ok is correct otherwise no need to change it

yosefe commented:
> imo, this change does not make sence if we use debug log returning ok is correct otherwise no need to change it

Indeed
1. how the issue is reproduced - which test/command?
2. According to PR desc, we should check for context release. if released - ucs_debug, return OK. otherwise the failure is real -> ucs_warn, return failure status.

rakhmets commented:
> > imo, this change does not make sence if we use debug log returning ok is correct otherwise no need to change it
> 
> Indeed
> 
> 1. how the issue is reproduced - which test/command?
> 2. According to PR desc, we should check for context release. if released - ucs_debug, return OK. otherwise the failure is real -> ucs_warn, return failure status.

1. The issue was found using the test from https://github.com/openucx/ucx/pull/10601. The code flow was described in this comment: https://github.com/openucx/ucx/pull/10606#discussion_r2030838703.
2. Yes, the proper fix is to store the information about the corresponding context in `uct_cuda_copy_alloc_handle_t`. It is required to store CUDA context handle, and CUDA context Id. Compare the returning value of `cuGetCtxId(uct_cuda_copy_alloc_handle_t::ctx)` to `uct_cuda_copy_alloc_handle_t::ctxId`. And do not call `cuMemFree` if `cuGetCtxId` fails, or the returning value doesn't match to the stored one.

rakhmets commented:
> > > imo, this change does not make sence if we use debug log returning ok is correct otherwise no need to change it
> > 
> > 
> > Indeed
> > 
> > 1. how the issue is reproduced - which test/command?
> > 2. According to PR desc, we should check for context release. if released - ucs_debug, return OK. otherwise the failure is real -> ucs_warn, return failure status.
> 
> 1. The issue was found using the test from [TEST/MPI: Added MPI+CUDA example. #10601](https://github.com/openucx/ucx/pull/10601). The code flow was described in this comment: [UCT/CUDA/CUDA_COPY: Changed log level in mem_free. #10606 (comment)](https://github.com/openucx/ucx/pull/10606#discussion_r2030838703).
> 2. Yes, the proper fix is to store the information about the corresponding context in `uct_cuda_copy_alloc_handle_t`. It is required to store CUDA context handle, and CUDA context Id. Compare the returning value of `cuGetCtxId(uct_cuda_copy_alloc_handle_t::ctx)` to `uct_cuda_copy_alloc_handle_t::ctxId`. And do not call `cuMemFree` if `cuGetCtxId` fails, or the returning value doesn't match to the stored one.
> 
> I just wanted to go with a simple fix for the release instead of the proper fix. Should we go with the proper one?

Force-pushed changes to update log level to `DIAG`.

rakhmets commented on src/uct/cuda/cuda_copy/cuda_copy_md.c:
```diff
@@ -616,7 +616,8 @@ static ucs_status_t uct_cuda_copy_mem_free(uct_md_h md, uct_mem_h memh)
         status = uct_cuda_copy_mem_release_fabric(alloc_handle);
     } else {
-        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemFree(alloc_handle->ptr));
+        UCT_CUDADRV_FUNC(cuMemFree(alloc_handle->ptr), UCS_LOG_LEVEL_DIAG);
```
The CUDA Driver API calls used in  uct_cuda_copy_mem_release_fabric do not required a CUDA context to be set, so we can keep them as is for now. However, errors should be replaced by warnings since it is a clean up flow. Not sure if it should be done in this PR, or can be postponed to be fixed in the next PR with other fixes/improvements/re-factoring.

yosefe replied:
Next PR
